Add Unified Push support

drizzt opened the pull request:
This simple implementation only triggers the networks to resume when an
event is received, so Telegram can get the events by itself.

This is not like the other push providers where the change information
is processed by PushListenerController directly, but it still saves battery
since the process disconnects from Telegram servers as soon as the
updates are processed.

I also reverted ["[TF][PUSH] remove one of the confusing toggles"](https://github.com/Telegram-FOSS-Team/Telegram-FOSS/commit/53a496326f632c594edb61b0aa7898f36fec77bf), since with UnifiedPush it's worthless to enable the pushService when UnifiedPush is used.

It fixes #577

Commit:
Revert "[TF][PUSH] remove one of the confusing toggles"

This reverts commit 53a496326f632c594edb61b0aa7898f36fec77bf.

Commit:
[TF][PUSH] add UnifiedPush settings in "Notifications and Sounds"

Commit:
Partially revert "[TF][PUSH] push service adjustments"

This reverts the NotificationsSettingsActivity.java part of commit
53a496326f632c594edb61b0aa7898f36fec77bf.

Commit:
[TF][PUSH] implemented a simple UnifiedPush support

This simple implementation only triggers the networks to resume when an
event is received, so Telegram can get the events by itself.

This is not like the other push providers where the change information
is processed by PushListenerController directly, but it still saves battery
since the process disconnects from Telegram servers as soon as the
updates are processed.

warioishere commented:
@drizzt 
hi drizzt, been follwing this since you came up with the fork, nice work, we are happy seeing this on Telegram
Just got some issues here and dont know where to post, as i dont find an issue area for mercurygram.
I just wanted to say that your first fork (telegram-foss-unified-push) worked well. then i switched to the new updated with rebanding to mercurygram.
Push doesnt work anymore on our own ntfy-server. If i can help you in anyway let me know, although im not a programmer

Greetings

drizzt commented:
Due to android limitations the battery optimization should be turn off, but you don't have battery usage if you turn off keep alive and background tasks.

p1gp1g commented on TMessagesProj/build.gradle:
```diff
@@ -33,6 +34,8 @@ dependencies {
     implementation 'com.google.code.gson:gson:2.10'
     implementation 'com.google.guava:guava:31.1-android'
 
+    implementation 'com.github.UnifiedPush:android-connector:2.1.1'
```
```suggestion
    implementation 'com.github.UnifiedPush:android-connector:2.2.0'
```

drizzt replied:
to do that it's necessary to change the java version used by the project or it won't work (currently JavaVersion.VERSION_1_8 is used). I'll try to change it and see if it still works

p1gp1g commented on TMessagesProj/src/main/java/org/telegram/messenger/PushListenerController.java:
```diff
@@ -1472,4 +1477,59 @@ public boolean hasServices() {
+                    SharedConfig.pushStringGetTimeStart = SystemClock.elapsedRealtime();
+                    SharedConfig.saveConfig();
+                    if (UnifiedPush.getDistributor(ApplicationLoader.applicationContext).isEmpty()) {
+                        List<String> distributors = UnifiedPush.getDistributors(ApplicationLoader.applicationContext, new ArrayList<>());
```
```suggestion
                        List<String> distributors = UnifiedPush.getDistributors(ApplicationLoader.applicationContext, new ArrayList<String>(Collections.singleton(UnifiedPush.FEATURE_BYTES_MESSAGE)));
```

It is required for webpush

drizzt replied:
yes, but I don't use webpush since telegram server still uses the draft encoding (https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-encryption-encoding-03) that sends the keys in headers and it requires a gateway + modifications of the code that complicates stuff a lot so I choose to just use SIMPLE push https://core.telegram.org/api/push-updates that is ONLY a put with version=NUMBER and so FEATURE_BYTE_MESSAGE is not needed

p1gp1g replied:
Ok, makes sens. All the distributors use the feature atm

p1gp1g commented on TMessagesProj/src/main/java/org/telegram/messenger/PushListenerController.java:
```diff
@@ -1472,4 +1477,59 @@ public boolean hasServices() {
+                    UnifiedPush.registerApp(
+                            ApplicationLoader.applicationContext,
+                            "default",
+                            new ArrayList<>(),
```
```suggestion
                            new ArrayList<String>(Collections.singleton(UnifiedPush.FEATURE_BYTES_MESSAGE))),
```

same

p1gp1g commented on TMessagesProj/src/main/java/org/telegram/messenger/UnifiedPushReceiver.java:
```diff
@@ -0,0 +1,83 @@
+                if (BuildVars.LOGS_ENABLED) {
+                    FileLog.d("UP START PROCESSING");
+                }
+                for (int a = 0; a < UserConfig.MAX_ACCOUNT_COUNT; a++) {
```
You can use the `instance` (https://github.com/Telegram-FOSS-Team/Telegram-FOSS/pull/715/files#diff-214ddbae5110be032e120290f658254085a490a2a5c1ca08e01542f3e95a9febR1520) to link to one account, to avoid fetching for all of them, but that's just an optimization

drizzt replied:
the point is that a notification can be for any user, since in registerDevice (https://core.telegram.org/method/account.registerDevice) you specify also other_uids that includes all the users (https://github.com/Telegram-FOSS-Team/Telegram-FOSS/blob/master/TMessagesProj/src/main/java/org/telegram/messenger/MessagesController.java#L12730) and so you can't just fetch the current user

p1gp1g replied:
ok, makes sens. It is still possible to have an endpoint per user if you want

p1gp1g commented on TMessagesProj/build.gradle:
```diff
@@ -5,6 +5,7 @@ apply plugin: 'com.android.library'
 repositories {
     mavenCentral()
     google()
+    maven { url 'https://www.jitpack.io' }
```
```suggestion
    maven {
            url "https://www.jitpack.io"
            content {
                includeModule 'com.github.UnifiedPush', 'android-connector'
            }
        }
```

To avoid other library to be fetched from jitpack

p1gp1g commented on TMessagesProj_App/build.gradle:
```diff
@@ -3,6 +3,7 @@ apply plugin: 'com.android.application'
 repositories {
     mavenCentral()
     google()
+    maven { url 'https://www.jitpack.io' }
```
```suggestion
    maven {
            url "https://www.jitpack.io"
            content {
                includeModule 'com.github.UnifiedPush', 'android-connector'
            }
        }
```

To avoid other library to be fetched from jitpack

drizzt commented:
if you agree with my objections I'll change the 3 remaining stuff and thank you for the review

Commit:
[TF][PUSH] avoid getting other libraries from jitpack

As suggested by @p1gp1g

Commit:
[TF][PUSH] update android-connector to 2.2.0

Since android-connector is built with Java 17 some other build system
modifications were needed.

This also means ./gradlew must be called using Java 17+.

thermatk commented:
looks awesome. Sorry for not writing sooner or reviewing, I'm caught in a personal situation. I have switched base to non-master so that it is still clean and I can release update on master asap, while this can be tackled in a slow manner :)

p1gp1g commented:
android-connector:2.3.0 uses jvmToolchain, it should work with java 1.8. You may have to add this to your build.gradle

```
kotlin {
    jvmToolchain(8)
}

drizzt commented:
> android-connector:2.3.0 uses jvmToolchain, it should work with java 1.8. You may have to add this to your build.gradle
> 
> ```
> kotlin {
>     jvmToolchain(8)
> }
> ```

I tried to do that, but it still needs a newer java:

> Task :TMessagesProj:compileReleaseJavaWithJavac
/home/tredaell/dev/sources/Mercurygram/TMessagesProj/src/main/java/org/telegram/messenger/UnifiedPushReceiver.java:9: error: cannot access MessagingReceiver
import org.unifiedpush.android.connector.MessagingReceiver;
                                        ^
  bad class file: /home/tredaell/.gradle/caches/transforms-3/d89406eb52297902fd6a4f25745d4ff1/transformed/jetified-android-connector-2.3.0-api.jar(/org/unifiedpush/android/connector/MessagingReceiver.class)
    class file has wrong version 61.0, should be 55.0

p1gp1g commented:
Indeed, it seems you need gradle 8+

```diff
diff --git a/TMessagesProj/build.gradle b/TMessagesProj/build.gradle
index 5f733bef..edff97a0 100644
--- a/TMessagesProj/build.gradle
+++ b/TMessagesProj/build.gradle
@@ -39,7 +39,7 @@ dependencies {
     implementation 'com.google.code.gson:gson:2.10'
     implementation 'com.google.guava:guava:31.1-android'
 
-    implementation 'com.github.UnifiedPush:android-connector:2.2.0'
+    implementation 'com.github.UnifiedPush:android-connector:2.3.1'
 
     implementation 'org.osmdroid:osmdroid-android:6.1.14'
 
@@ -84,8 +84,8 @@ android {
     }
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_17
-        targetCompatibility JavaVersion.VERSION_17
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
 
         coreLibraryDesugaringEnabled true
     }
@@ -193,6 +193,8 @@ android {
             buildConfigField "boolean", "BUILD_HOST_IS_WINDOWS", isWindows
         }
     }
+
+    namespace = "org.telegram.messenger"
 }
 
 
diff --git a/TMessagesProj_App/build.gradle b/TMessagesProj_App/build.gradle
index 7c2b7305..7d762e9e 100644
--- a/TMessagesProj_App/build.gradle
+++ b/TMessagesProj_App/build.gradle
@@ -44,8 +44,8 @@ android {
     }
 
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_17
-        targetCompatibility JavaVersion.VERSION_17
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
 
         coreLibraryDesugaringEnabled true
     }
@@ -181,4 +181,6 @@ android {
     lintOptions {
         checkReleaseBuilds false
     }
+
+    namespace = "org.telegram.messenger.regular"
 }
\ No newline at end of file
diff --git a/build.gradle b/build.gradle
index f48ee03c..addb3b69 100644
--- a/build.gradle
+++ b/build.gradle
@@ -5,9 +5,7 @@ buildscript {
         google()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.0.3'
-        // This is needed since android-connector:2.2.0 is built with Java 17
-        classpath 'com.android.tools:r8:8.2.33'
+        classpath 'com.android.tools.build:gradle:8.2.0'
     }
 }
 repositories {
diff --git a/gradle/wrapper/gradle-wrapper.properties b/gradle/wrapper/gradle-wrapper.properties
index ffed3a25..b26d0324 100644
--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,6 @@
+#Thu Dec 28 14:56:03 CET 2023
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.2-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.2-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

p1gp1g commented:
You shouldn't even have to upgrade gradle with connector:2.4.0.

lone-cloud commented:
I'd be interested to know what's holding this PR up? The current FOSS implementation is too much of a battery hog to use imo. I did try mercurygram, but push notifications don't work for me there regardless of which UP distribution I try (tested with ntfy and conversations) and regardless of the ntfy server (tested ntfy with several including drizzt's personal server). Note that I'm running ntfy for 3 other apps on my phone and mercurygram is the only app that doesn't work and I've spent hours debugging and testing it.
I'm curious if anyone knows how to create an apk to test this PR? I'm guessing the only way is to set up the android studio dev environment and to build it locally?

Integral-Tech commented:
Any progress on this PR?

ildar commented:
Can anyone confirm it works? I still find it hard to believe it does.

warioishere commented:
it works, but sometimes notifications are heavily delayed. Using Mercurygram latest and own ntfy server and ntfy client on android

> Can anyone confirm it works? I still find it hard to believe it does.

ildar commented:
interesting! the delay is on the T-server's side?
